src/implicit.rs

```diff
@@ -477,37 +481,30 @@ struct BracketPair {
 /// text source.
 ///
 /// <https://www.unicode.org/reports/tr9/#BD16>
-fn identify_bracket_pairs<D: BidiDataSource>(
-    text: &str,
+fn identify_bracket_pairs<'a, T: TextSource<'a> + ?Sized, D: BidiDataSource>(
+    text: &'a T,
     data_source: &D,
     run_sequence: &IsolatingRunSequence,
     original_classes: &[BidiClass],
-) -> Vec<BracketPair> {
+) -> Vec<BracketPair>
+where
+    <T as TextSource<'a>>::CharIter: Iterator<Item = char>,
+{
     let mut ret = vec![];
     let mut stack = vec![];
 
     for (run_index, level_run) in run_sequence.runs.iter().enumerate() {
-        let slice = if let Some(slice) = text.get(level_run.clone()) {
-            slice
-        } else {
-            #[cfg(feature = "std")]
-            std::debug_assert!(
-                false,
-                "Found broken indices in level run: found indices {}..{} for string of length {}",
-                level_run.start,
-                level_run.end,
-                text.len()
-            );
-            return ret;
-        };
-
-        for (i, ch) in slice.char_indices() {
-            let actual_index = level_run.start + i;
+        let mut actual_index = level_run.start;
+
+        for ch in text.subrange(level_run.clone()).chars() {
+            let char_len = T::char_len(ch);
```

Comment:
nit: let's use char_indices here instead of recomputing char_len. there's a chance the optimizer will figure this out, I'd rather not take that chance.

Again, the old code had the right structure: run char_indices and add to `level_run.start`.

Reply:
Done in [7866637](https://github.com/servo/unicode-bidi/pull/109/commits/786663784fc5f5c8b700c9b2fd7ebcb4227aeb04)